pkg/controller/clusterapimachineset/machineset_controller.go

```diff
@@ -0,0 +1,293 @@
+	key, err := controller.KeyFunc(ms)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", ms, err))
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+// enqueueAfter will enqueue a cluster after the provided amount of time.
+func (c *Controller) enqueueAfter(cluster *clustopv1.Cluster, after time.Duration) {
+	key, err := controller.KeyFunc(cluster)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", cluster, err))
+		return
+	}
+
+	c.queue.AddAfter(key, after)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("Sync %q failed with %v", key, err))
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+// syncCluster will sync the cluster with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (c *Controller) syncMachineSet(key string) error {
+	startTime := time.Now()
+	c.logger.WithField("key", key).Debug("syncing machineset")
+	defer func() {
+		c.logger.WithFields(log.Fields{
+			"key":      key,
+			"duration": time.Now().Sub(startTime),
+		}).Debug("finished syncing machineset")
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	ms, err := c.machineSetsLister.MachineSets(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		c.logger.WithField("key", key).Debug("machineset has been deleted")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	msLog := colog.WithClusterAPIMachineSet(c.logger, ms)
+	msLog.Debugf("found machineset")
+
+	// Lookup the cluster for this machine set:
+	clusterName, ok := ms.Labels[clusterNameLabel]
+	if !ok {
+		return fmt.Errorf("no cluster label set on machine set")
+	}
+	cluster, err := c.clustersLister.Clusters(namespace).Get(clusterName)
+	if errors.IsNotFound(err) {
+		msLog.WithField("cluster", clusterName).Errorf("no cluster found for machineset")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	clusterSpec, err := controller.ClusterSpecFromClusterAPI(cluster)
+	if err != nil {
+		return err
+	}
+
+	clusterVersion, err := c.clustopClient.ClusteroperatorV1alpha1().ClusterVersions(clusterSpec.ClusterVersionRef.Namespace).Get(clusterSpec.ClusterVersionRef.Name, metav1.GetOptions{})
+	if err != nil {
+		msLog.WithField("cluster version does not exist: %v", clusterSpec.ClusterVersionRef)
+		return err
+	}
+
+	machineSpec, err := controller.PopulateMachineSpec(ms.Spec.Template.Spec, clusterSpec, clusterVersion, msLog)
```

Comment:
I don't understand why we have this controller at all. Presumably, this is running in the root cluster, where the user is creating the MachineSet objects. We should not have a controller that is modifying user-created objects. The whole point of the capi-machine controller is to update the template for the Machine objects, that cluster-operator (via cluster-api) owns.

If we stop trying to write the Cluster and ClusterVersion information into the MachineSet, then we won't have a problem with trying to store user-specified and cluster-derived defaults in the same field as detailed in my comment in `PopulateMachineSpec`.

Reply:
The controller exists to make sure the spec template is updated before we send down to the target cluster (soon) and it starts to act on it, as we don't intend to have the capi-machine controller running remotely and there are no cluster versions defined. I expected we would do similar for MachineDeployment. 

We need some place to pass that data, and ideally do some calculation logic that remains under the root clusters control. If all of MachineTemplateSpec providerConfig is off limits I'm not sure how we do that.

Reply:
I'm not sure why we wouldn't fill out the template at the time that we are sending the data to the remote cluster. I don't see why we need to mutate our root cluster MachineSet objects to satisfy the information being in the remote Machine templates.

Reply:
We could but then we've got different representations of the same thing in the root and the target which strikes me as quite unusual, and if we get into syncing or ever supporting changes made remotely, we'll have some problems to deal with. Possible that can be reconciled but it's something we should watch out for. I can back this controller up and pull it out of this PR though.